### ISSUE TITLE
Zalenium unable to allocate sessions to nodes

### ISSUE THREAD

**Issue by kallekrantz** [zalando/zalenium#833]
During the past weeks, we've run into problems running our Zalenium grid where nodes are not disposed and remain completely unused (but unable to accept new sessions). This then blocks all further allocations of selenium sessions, since new nodes cannot be started due to the maximum number of containers having been created.

This issue has persisted across a few upgrades and happens roughly a day or so after a restart of the grid itself.

Zalenium Image Version(s):  
`3.141.59`
Docker Version:  
`18.06.1-ce`
If using docker-compose, version:  
`1.22.0`
OS: 
`CentOS Linux release 7`
Docker Command to start Zalenium:  
**docker-compose file used to start the grid.**
```
services:
  zalenium:
    image: "dosel/zalenium"
    container_name: zalenium
    hostname: zalenium
    tty: true
    volumes:
      - /var/lib/docker/volumes/zalenium_videos:/home/seluser/videos
      - /var/run/docker.sock:/var/run/docker.sock
      - /usr/bin/docker:/usr/bin/docker
    ports:
      - 4444:4444
    command: >
      start --desiredContainers 5
            --maxDockerSeleniumContainers 20
            --screenWidth 1920 --screenHeight 1080
            --videoRecordingEnabled true
            --retentionPeriod 3
```
## Expected Behavior -

Nodes are created and disposed of normally, i.e nodes are not stuck unused

## Actual Behavior -
The maximum number of nodes up to the --maxDockerSeleniumContainers are created, followed by the grid freezing and reporting the following stack trace on every new allocation request:

```
12:43:23.400 [qtp1591916281-24527] INFO  o.o.g.w.s.handler.RequestHandler - Error forwarding the new session Request timed out waiting for a node to become available.
java.util.concurrent.TimeoutException: Request timed out waiting for a node to become available.
        at org.openqa.grid.web.servlet.handler.RequestHandler.waitForSessionBound(RequestHandler.java:197)
        at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:112)
        at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:85)
        at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:69)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:865)
        at org.seleniumhq.jetty9.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
        at io.prometheus.client.filter.MetricsFilter.doFilter(MetricsFilter.java:170)
        at org.seleniumhq.jetty9.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:533)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
        at org.seleniumhq.jetty9.security.SecurityHandler.handle(SecurityHandler.java:548)
        at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1340)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:473)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:1564)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1242)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
        at org.seleniumhq.jetty9.server.handler.StatisticsHandler.handle(StatisticsHandler.java:174)
        at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.seleniumhq.jetty9.server.Server.handle(Server.java:503)
        at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:364)
        at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:260)
        at org.seleniumhq.jetty9.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
        at org.seleniumhq.jetty9.io.FillInterest.fillable(FillInterest.java:103)
        at org.seleniumhq.jetty9.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
        at org.seleniumhq.jetty9.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
        at java.lang.Thread.run(Thread.java:748)
```
This is the state of the console. The live vnc-preview shows them all being completely unused (but are still marked as active with tests).
![image](https://user-images.githubusercontent.com/2708253/51182214-bd9c2500-18cd-11e9-9864-1004f99ff72b.png)

I have a hard time parsing the logs themselves, but from what I can tell from speaking to our internal test owners it seems to happen the morning after about 300 tests ran (5 in parallel), or after the execution of 55 tests (5 in parallel).

Just let me know what other information or debugging steps I can perform to assist!

**Comment by diemol** [comment]
Hi @kallekrantz,

What is the exact version of Zalenium that you are using? Recently we released something that should mitigate that behaviour.

**Comment by kallekrantz** [comment]
Hi @diemol!

Thanks for the response! I have a hard-time finding out exactly what version I've got running, as it seems as though version isn't being printed on startup. I did update last week, after noticing a new version having been released:

```
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS                  PORTS                              NAMES
1b229b548ca8        dosel/zalenium           "entry.sh start --de…"   46 hours ago        Up 22 hours (healthy)   0.0.0.0:4444->4444/tcp, 4445/tcp   zalenium
```

```
REPOSITORY                     TAG                 IMAGE ID            CREATED             SIZE
dosel/zalenium                 latest              ac8ee5dfa92d        5 days ago          527MB
```
This is the same image that the currently running container is executing, `docker inspect 1b229b548ca8`:
```
$ docker inspect 1b229b548ca8
{
...
        "Image": "sha256:ac8ee5dfa92d45a520107ea18a94ea25c5310282330d499027359109c2a13790",
...
}
```
Also tried to pull a new version, it seems to think I'm running the latest one at least:
```
$ docker pull dosel/zalenium
Using default tag: latest
latest: Pulling from dosel/zalenium
Digest: sha256:7fea4d49b25ddb074adae8c3b82917f2aa5494a84b92bcbfd8f07ff201c356a2
Status: Image is up to date for dosel/zalenium:latest
```

**Comment by ashkb** [comment]
I am seeing the same issue on Zalenium 3.141.59f. Attached is the Zalenium log file.
[ZaleniumErrorLog.txt](https://github.com/zalando/zalenium/files/2766540/ZaleniumErrorLog.txt)

I only run into this issue when if I set the maxTestSessions to 1, resulting in containers getting started and torn down continuously as the tests runs.

**Comment by echeren** [comment]
We are also experiencing the same problem, our environment is running:

DOCKER | 18.09.0
UBUNTU_FLAVOR | xenial
ZAL_VER | zalenium-3.141.59f

**Comment by diemol** [comment]
OK, we will have a look in the next couple of days based on the comments. Thanks for reporting this.

**Comment by shyamkunti** [comment]
We are facing same issue with the latest version of zalenium. I am using below curl to start zalenium and after 20 tests I am getting Error forwarding the new session Request timed out waiting for a node to become available.

curl -sSL https://raw.githubusercontent.com/dosel/t/i/p | bash -s start --videoRecordingEnabled true --maxDockerSeleniumContainers 20

**Comment by megolee** [comment]
I have the same issue in my environment, here is the log  
https://gist.githubusercontent.com/megolee/36e4354700d5120afb84688da1bbfb8d/raw/edaa8f2bdfc90df63fddbb3478639e219b9fbab4/Zalenium%2520Logs

**Comment by skunti** [comment]
We temporarily resolved this by pulling an old container by below change

`curl -sSL https://raw.githubusercontent.com/dosel/t/i/p | bash -s start 3.14.0g --videoRecordingEnabled true --maxDockerSeleniumContainers 20`

**Comment by megolee** [comment]
But I still have the same problem with 3.14.0g, all the selenium node stuck and not release after use. This make all the next cases fail.

**Comment by diemol** [comment]
Sorry, I haven't had time to do anything related to this. New job and new baby are taking most of my time.

If anyone has time to track down the root cause or perhaps an area of code that should be double checked, that would be a helpful start.

**Comment by kallekrantz** [comment]
I actually went over the logs from some of the recent freezes, and I noticed that the behaviour isn't really the same anymore. We still see the timeouts from the client side, but now the logs are full of the following messages:

```
13:54:41.213 [RemoteProxy CleanUpThread for http://172.20.0.6:41150] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 9fc23abdde9b1bc81cdd64762802a125 has TIMED OUT due to client inactivity and will be released.
13:54:41.215 [RemoteProxy CleanUpThread for http://172.20.0.10:41135] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key e11e795f4268f96a258009134cb26979 has TIMED OUT due to client inactivity and will be released.
13:54:41.215 [RemoteProxy CleanUpThread for http://172.20.0.18:41142] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 0c5d105fb47934411a280a53ae7eaecd
13:54:41.215 [RemoteProxy CleanUpThread for http://172.20.0.6:41150] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 9fc23abdde9b1bc81cdd64762802a125
13:54:41.217 [RemoteProxy CleanUpThread for http://172.20.0.10:41135] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key e11e795f4268f96a258009134cb26979
13:54:42.734 [RemoteProxy CleanUpThread for http://172.20.0.11:41136] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 5391d798e473553ead4cf13b8bed2181 has TIMED OUT due to client inactivity and will be released.
13:54:42.734 [RemoteProxy CleanUpThread for http://172.20.0.14:41138] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 89a21c93ce5b87eadc1fc06c0325baef has TIMED OUT due to client inactivity and will be released.
13:54:42.734 [RemoteProxy CleanUpThread for http://172.20.0.17:41141] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 14d9aec4d0df6901aa73bd23bd23cc5d has TIMED OUT due to client inactivity and will be released.
13:54:42.735 [RemoteProxy CleanUpThread for http://172.20.0.19:41143] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key ce12b717185d9ee0756db4269096a4af has TIMED OUT due to client inactivity and will be released.
13:54:42.735 [RemoteProxy CleanUpThread for http://172.20.0.13:41137] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key fc3b8fdc5e53842687641e3a5d819af6 has TIMED OUT due to client inactivity and will be released.
13:54:42.735 [RemoteProxy CleanUpThread for http://172.20.0.16:41140] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 93c0e0f44f538a3a9908729c023ffacd has TIMED OUT due to client inactivity and will be released.
13:54:42.736 [RemoteProxy CleanUpThread for http://172.20.0.11:41136] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 5391d798e473553ead4cf13b8bed2181
13:54:42.736 [RemoteProxy CleanUpThread for http://172.20.0.14:41138] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 89a21c93ce5b87eadc1fc06c0325baef
13:54:42.737 [RemoteProxy CleanUpThread for http://172.20.0.21:41145] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 8487ee2b42422af1455a4ec10c67a8a2 has TIMED OUT due to client inactivity and will be released.
13:54:42.737 [RemoteProxy CleanUpThread for http://172.20.0.22:41147] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 74e4b7bbb59812b605dfa38de9169754 has TIMED OUT due to client inactivity and will be released.
13:54:42.737 [RemoteProxy CleanUpThread for http://172.20.0.17:41141] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 14d9aec4d0df6901aa73bd23bd23cc5d
13:54:42.737 [RemoteProxy CleanUpThread for http://172.20.0.13:41137] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key fc3b8fdc5e53842687641e3a5d819af6
13:54:42.737 [RemoteProxy CleanUpThread for http://172.20.0.16:41140] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 93c0e0f44f538a3a9908729c023ffacd
13:54:42.738 [RemoteProxy CleanUpThread for http://172.20.0.19:41143] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key ce12b717185d9ee0756db4269096a4af
13:54:42.738 [RemoteProxy CleanUpThread for http://172.20.0.21:41145] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 8487ee2b42422af1455a4ec10c67a8a2
13:54:42.738 [RemoteProxy CleanUpThread for http://172.20.0.22:41147] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 74e4b7bbb59812b605dfa38de9169754
```

I tried looking through the Selenium source code to find where that's being used, but could only find what essentially amounted to a POST call towards the specific node. And since Zalenium seems quite good about actually removing the containers after a restart, I can't get logs from the nodes.

This doesn't really help anything as of now or by itself, but I'm gonna try to dig deeper into a running docker container the next time it happens to see if I can get some additional logging.

**Comment by kallekrantz** [comment]
I ran some tests using https://github.com/kallekrantz/zalenium-test and I was able to reproduce similar behaviours where the grid no longer accepted connections. But for everytime the cause seemed slightly different. 

First error was due to Zalenium simply freezing, so I added to same autohealing mechanism that we run in our dev environment. Then another fault appeared, that we couldn't allocate more memory on the Java Heap - causing no more tests to be able to be scheduled. Both of the above happened after ~3 hours of continuous load.

Gonna continue experimenting however, will see if I can get the above behaviour reproduced.

PS. Noticed something interesting when doing some experiments, if the individual selenium containers are started at the same time as the grid is shutting down. they can be orphaned and won't shut down. They do however appear to be pruned upon the the next start of Zalenium, so this is unlikely to be the cause of this bug.

**Comment by diemol** [comment]
This seems to be the reason https://github.com/SeleniumHQ/selenium/issues/6918
A fix was merged to master but we need to figure out how to bring that code here.

**Comment by aarkanov** [comment]
Is there any update on this? I've been trying to use zalenium but running into issues where 

```
14:04:49.471 [qtp1601935322-32] INFO  o.o.g.w.s.handler.RequestHandler - Got a request to create a new session: Capabilities {applicationCacheEnabled: false, browserName: chrome, goog:chromeOptions: {args: [--start-maximized], extensions: []}, name: NavigationTests.ID01_C_Curr..., version: }
14:04:49.471 [qtp1601935322-32] INFO  d.z.e.z.registry.ZaleniumRegistry - Adding sessionRequest for {applicationCacheEnabled=false, name=NavigationTests.ID01_C_CurrencyProfiles_HomePageNavigation.i_MyrCurrencyProfilesLinks, browserName=chrome, goog:chromeOptions={args=[--start-maximized], extensions=[]}, version=}
14:04:52.802 [qtp1601935322-31] INFO  o.o.g.w.s.handler.RequestHandler - Error forwarding the new session Request timed out waiting for a node to become available.
java.util.concurrent.TimeoutException: Request timed out waiting for a node to become available.
	at org.openqa.grid.web.servlet.handler.RequestHandler.waitForSessionBound(RequestHandler.java:197)
	at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:112)
	at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:85)
	at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:69)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.seleniumhq.jetty9.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
	at io.prometheus.client.filter.MetricsFilter.doFilter(MetricsFilter.java:170)
	at org.seleniumhq.jetty9.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:533)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.seleniumhq.jetty9.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1340)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1242)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.seleniumhq.jetty9.server.handler.StatisticsHandler.handle(StatisticsHandler.java:174)
	at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.seleniumhq.jetty9.server.Server.handle(Server.java:503)
	at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:364)
	at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.seleniumhq.jetty9.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.seleniumhq.jetty9.io.FillInterest.fillable(FillInterest.java:103)
	at org.seleniumhq.jetty9.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.seleniumhq.jetty9.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
	at java.lang.Thread.run(Thread.java:748)

```

And then Zalenium can't create more machines and my tests just hang. I'm running it on a t3.xlarge ec2 instance with 5 containers
Thank you any help would be appreciated.

**Comment by skunti** [comment]
Use old revision just like mentioned above until issue is fixed.
curl -sSL https://raw.githubusercontent.com/dosel/t/i/p | bash -s start 3.14.0g --videoRecordingEnabled true --maxDockerSeleniumContainers 20

**Comment by aarkanov** [comment]
Thank you @skunti how do I tell it where to record the videos using the curl command? so that it doesn't record it in a default space?

**Comment by Pratish-minted** [comment]
i am getting the same error even though using dosel/zalenium:3.14.0g 
```17:53:04    [testng] Unable to create driver instance in 1st attempt with retry timeout of 0 seconds. You can check/set value of 'driver.init.retry.timeout' appropriately to set retry timeout on driver initialization failure.Unable to Create Driver Instance Error forwarding the new session Request timed out waiting for a node to become available.
17:53:04    [testng] Command duration or timeout: 600.01 seconds```

**Comment by Pratish89** [comment]
i am getting the same error even though i used dosel/zalenium:3.14.0g 

` Unable to create driver instance in 1st attempt with retry timeout of 0 seconds. You can check/set value of 'driver.init.retry.timeout' appropriately to set retry timeout on driver initialization failure.Unable to Create Driver Instance Error forwarding the new session Request timed out waiting for a node to become available.`

**Comment by hajtomoto** [comment]
we had exactly same issue in our place too. For days, no one knew how to resolve it so we end-up giving up on Zalenium and reverted to single instance of Selenium. Won't be using Zalenium until this issue is resolved :(

**Comment by Arun-AG** [comment]
08:55:53.196 [qtp79290250-2939] INFO  o.o.g.w.s.handler.RequestHandler - Error forwarding the new session Request timed out waiting for a node to become available.
java.util.concurrent.TimeoutException: Request timed out waiting for a node to become available.
        at org.openqa.grid.web.servlet.handler.RequestHandler.waitForSessionBound(RequestHandler.java:197)
        at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:112)
        at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:85)
        at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:69)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:865)
        at org.seleniumhq.jetty9.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
        at io.prometheus.client.filter.MetricsFilter.doFilter(MetricsFilter.java:170)
        at org.seleniumhq.jetty9.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:533)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
        at org.seleniumhq.jetty9.security.SecurityHandler.handle(SecurityHandler.java:548)
        at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1340)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:473)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:1564)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1242)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
        at org.seleniumhq.jetty9.server.handler.StatisticsHandler.handle(StatisticsHandler.java:174)
        at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.seleniumhq.jetty9.server.Server.handle(Server.java:503)
        at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:364)
        at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:260)
        at org.seleniumhq.jetty9.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
        at org.seleniumhq.jetty9.io.FillInterest.fillable(FillInterest.java:103)
        at org.seleniumhq.jetty9.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
        at org.seleniumhq.jetty9.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
        at java.lang.Thread.run(Thread.java:748)
08:55:53.215 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:55:53.218 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:55:58.219 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:55:58.222 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:03.222 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:03.225 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:08.225 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:08.228 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:09.415 [qtp79290250-2994] INFO  o.o.g.w.s.handler.RequestHandler - Got a request to create a new session: Capabilities {acceptSslCerts: true, browserName: chrome, browser_version: 59, browserstack.debug: true, browserstack.local: true, idleTimeout: 300, marionette: true, os: Windows, os_version: 7}
08:56:09.415 [qtp79290250-2994] INFO  d.z.e.z.registry.ZaleniumRegistry - Adding sessionRequest for {browserstack.local=true, acceptSslCerts=true, os=Windows, marionette=true, idleTimeout=300, os_version=7, browserName=chrome, browser_version=59, browserstack.debug=true}
08:56:13.229 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:13.232 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:18.232 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:18.235 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:23.235 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:23.238 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:28.239 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:28.242 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:33.242 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:33.245 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:38.245 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:38.248 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:43.249 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.
08:56:43.252 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.g.s.proxy.DefaultRemoteProxy - Error releasing the resources on timeout for session ext. key 559efca51fb44bfe0f37728deacb630e
08:56:48.252 [RemoteProxy CleanUpThread for http://172.17.0.3:40000] WARN  o.o.grid.internal.BaseRemoteProxy - session ext. key 559efca51fb44bfe0f37728deacb630e has TIMED OUT due to client inactivity and will be released.

**Comment by Arun-AG** [comment]
Same issue.  Getting this issue. Could not Find the Exact reason for it

**Comment by AmitMaheshwari-minted** [comment]
Selenoid may be an alternative solution in the meantime using the same set of computing resources.

**Comment by jaasal** [comment]
Any news on this? I would consider this as blocker issue.

**Comment by evertsc** [comment]
I'm also having the same problem with the latest versions. Would be great if they can be solved.

Went back to dosel/zalenium 3.8.1i with elgalu/selenium 3.8.1-p13 where I don't have these problems.

**Comment by diemol** [comment]
Fixed via #960, will be released soon.

**Comment by Pratish89** [comment]
@diemol  could you please update here once this fix gets released tag? 
Thanks.

**Comment by diemol** [comment]
It was released earlier today.

**Comment by dunjoye4real** [comment]
@diemol  I have the same problem. Kindly update the Fix here.
BR

**Comment by diemol** [comment]
@dunjoye4real it was already released, if you have still the issue, please provide a way to reproduce it in a new issue.

**Comment by Pratish89** [comment]
@diemol thanks will check this out with latest image and update here.

**Comment by rootux** [comment]
I'm still having this issue on `latest version` released on `12/9/2019` - I guess this is `3.141.59w`
Freezing nodes - Seems to be allocated but actually just being frozen while I get timeouts
<img width="1335" alt="Screen Shot 2020-01-13 at 11 10 23" src="https://user-images.githubusercontent.com/1032526/72243776-71291980-35f5-11ea-8a19-16ba30145696.png">

```
docker --version
Docker version 18.09.9-ce, build 039a7df
```

**Comment by labordoc** [comment]
Hi @diemol:

We recently started using Zalenium and we are facing the same issue reported in this thread:  for some executions everything goes fine (all tests run and nodes get disclosed and available again), but in other cases, there are some nodes that do not get disclosed and become unavailable, so tests do not use those nodes anymore, and the only solution we saw is to restart Zalenium, so all nodes get available again.

After reading all previous comments, I understand the issue has not been fixed. @diemol : do you know if there is a way to detect automatically the "unavailable" nodes and restart them individually?

Thanks!

**Comment by diemol** [comment]
I see some comments here and in #1162, but no way to reproduce it, therefor I cannot really troubleshoot and help. Sorry.
